pysilcam/tracking/track.py

```diff
@@ -0,0 +1,417 @@
+    :return:
+    '''
+
+    print('* make_output_files_for_giffing')
+
+    os.makedirs(outputdir, exist_ok=True)
+
+    u = np.unique(data['t2'])
+
+    for timestamp in u:
+        print('making tracks for', timestamp)
+        tmptracks = data[pd.to_datetime(data['t2']) == pd.to_datetime(timestamp)]
+        print('len(tmptracks)', len(tmptracks))
+        if len(tmptracks) == 0:
+            continue
+
+        try:
+            im = im_from_timestamp(pd.to_datetime(timestamp), rawdatapath)
+            im = np.uint8(np.min(im, axis=2))
+        except:
+            print('could not load image from:', str(tmptracks.iloc[0]['t2']))
+            continue
+
+        plt.figure(figsize=(7, 10))
+        r, c = np.shape(im)
+        plt.imshow(rotate(explode_contrast(np.uint8(im)),
+                          270, resize=True),
+                   cmap='gray',
+                   extent=[0, r * PIX_SIZE / 1000,
+                           0, c * PIX_SIZE / 1000])
+        plt.title(str(timestamp))
+
+        subset = data[data['t2'] <= pd.to_datetime(timestamp)]
+
+        for p in subset['ParticleName'].values:
+            this_particle = subset[subset['ParticleName'] == p]
+            if len(this_particle) == 0:
+                continue
+
+            t_arr = min(this_particle['t1'])
+            t_dep = max(this_particle['t2'])
+            x_arrival = this_particle[this_particle['t1'] == t_arr]['x1'].values[0]
+            x_departure = this_particle[this_particle['t2'] == t_dep]['x2'].values[0]
+            linex = np.float64([x_arrival, x_departure])
+            y_arrival = this_particle[this_particle['t1'] == t_arr]['y1'].values[0]
+            y_departure = this_particle[this_particle['t2'] == t_dep]['y2'].values[0]
+            liney = np.float64([y_arrival, y_departure])
+
+            speed = calculate_speed(x_arrival, x_departure, y_arrival, y_departure, t_arr, t_dep, PIX_SIZE)
+
+            linex *= PIX_SIZE / 1000
+            liney *= PIX_SIZE / 1000
+            linex = c * PIX_SIZE / 1000 - linex
+            liney = r * PIX_SIZE / 1000 - liney
+            n_tracks = len(this_particle)  # should be the same as max(this_particle['n-tracks'])
+            plot_color = 'r-'
+            if n_tracks > track_length_limit:
+                plot_color = 'g-'
+
+                plt.text(liney[0], linex[0],
+                         (p + '\n{:0.2f}mm {:0.2f}mm/s'.format(this_particle['length'].values[0] * PIX_SIZE / 1000,
+                                                               speed * 10)),
+                         fontsize=4, color='g')
+
+            plt.plot(liney, linex, plot_color, linewidth=1)
+
+        plt.xlabel('mm')
+        plt.ylabel('mm')
+        plt.gca().invert_yaxis()
+
+        name = pd.to_datetime(timestamp).strftime('D%Y%m%dT%H%M%S.%f')
+        print('saving', os.path.join(outputdir, name + '-tr.png'))
+        plt.savefig(os.path.join(outputdir, name + '-tr.png'), dpi=300, bbox_inches='tight')
```

Comment:
F405: 'os' may be undefined, or defined from star imports: pysilcam.tracking.silcamtracker  <!-- Automatically posted by Lintly -->